common/hashing_test.cpp

```diff
@@ -0,0 +1,589 @@
+    CARBON_DCHECK(v == hashes[bits_and_indices.size()].v);
+    auto hash_bits = (static_cast<uint64_t>(hash) & BitMask) >> BitShift;
+    bits_and_indices.push_back(
+        {static_cast<uint32_t>(hash_bits), bits_and_indices.size()});
+  }
+  std::sort(
+      bits_and_indices.begin(), bits_and_indices.end(),
+      [](const auto& lhs, const auto& rhs) { return lhs.first < rhs.first; });
+
+  uint32_t prev_hash_bits = bits_and_indices[0].first;
+  int prev_index = bits_and_indices[0].second;
+  bool in_collision = false;
+  int total = 0;
+  int distinct_collisions = 0;
+  for (const auto& [hash_bits, hash_index] :
+       llvm::ArrayRef(bits_and_indices).slice(1)) {
+    if (hash_bits != prev_hash_bits) {
+      prev_hash_bits = hash_bits;
+      prev_index = hash_index;
+      in_collision = false;
+      continue;
+    }
+
+    ++total;
+    if (in_collision) {
+      ++collision_counts.back();
+      collision_map[hash_index] = collision_counts.size() - 1;
+      continue;
+    }
+    in_collision = true;
+    collision_map[hash_index] = collision_counts.size();
+    collision_counts.push_back(1);
+
+    ++distinct_collisions;
+    if (0 && distinct_collisions < 10) {
+      llvm::errs() << "Hash mask " << llvm::formatv("{0:x16}", BitMask)
+                   << " collision: " << hashes[prev_index] << " vs. "
+                   << hashes[hash_index] << "\n";
+    }
+  }
+
+  // Sort by collisions.
+  std::sort(bits_and_indices.begin(), bits_and_indices.end(),
+            [&](const auto& lhs, const auto& rhs) {
+              return collision_counts[collision_map[lhs.second]] <
+                     collision_counts[collision_map[rhs.second]];
+            });
+
+  int median = collision_counts
+      [collision_map[bits_and_indices[bits_and_indices.size() / 2].second]];
+  int max = collision_counts.back();
+  return {.total = total, .median_per_hash = median, .max_per_hash = max};
+}
+
+template <int N>
+auto AllByteStringsHashedAndSorted() {
+  static_assert(N < 5, "Can only generate all 4-byte strings or shorter.");
+
+  llvm::SmallVector<HashedString> hashes;
+  int64_t count = 1LL << (N * 8);
+  for (int64_t i : llvm::seq(count)) {
+    uint8_t bytes[N];
+    for (int j : llvm::seq(N)) {
+      bytes[j] = (static_cast<uint64_t>(i) >> (8 * j)) & 0xff;
+    }
+    std::string s(std::begin(bytes), std::end(bytes));
+    hashes.push_back({HashValue(s, TestSeed), s});
+  }
+
+  std::sort(hashes.begin(), hashes.end(),
+            [](const HashedString& lhs, const HashedString& rhs) {
+              return static_cast<uint64_t>(lhs.hash) <
+                     static_cast<uint64_t>(rhs.hash);
+            });
+
+  return hashes;
+}
+
+TEST(HashingTest, Collisions1ByteSized) {
+  auto hashes_storage = AllByteStringsHashedAndSorted<1>();
+  auto hashes = llvm::ArrayRef(hashes_storage);
+
+  HashCode prev_hash = hashes[0].hash;
+  llvm::StringRef prev_s = hashes[0].v;
+  for (const auto& [hash, s] : llvm::ArrayRef(hashes).slice(1)) {
```

Comment:
```suggestion
  for (const auto& [hash, s] : hashes.slice(1)) {
```
(Also in the next function)